pytorch_blade/torch_blade/onnx_backends/backend_testbed.py

```diff
@@ -38,17 +38,37 @@ def is_tensor(val):
         )
         return all_concrete_inputs and all_tensor_outputs
 
+    def _patch_inputs_scalar_types(self, graph, scalar_types):
+        """Add scalar type and dim to plain TensorType."""
+        assert len(list(graph.inputs())) == len(scalar_types)
+        for inp, scalar_type in zip(graph.inputs(), scalar_types):
+            inp_type = inp.type()
+            if scalar_type and inp_type.isSubtypeOf(torch._C.TensorType.get()):
+                new_type = tools.tensor_type_from_scalar(scalar_type)
+                inp.setType(new_type)
+
+    def _record_inputs_scalar_types(self, graph):
+        scalar_inputs = []
+        for inp in graph.inputs():
+            if inp.type().isSubtypeOf(torch._C.NumberType.get()):
+                scalar_inputs.append(inp.type())
+            else:
+                scalar_inputs.append(None)
+        return scalar_inputs
+
     def __call__(self):
         try:
             graph = self._graph
             # use mannual rules to filter the graph
             if not onnx_lower_guard.check_graph_with_rules(graph):
                 return False
+            scalar_types = self._record_inputs_scalar_types(graph)
```

Comment:
Why should these scalar exists, since we had forbidden scalar as subgraph inputs without ingore_device=True, as is L379.

Reply:
I think we can put some declarations and examples here.

Reply:
They're indirect inputs introduced by recursive `_appendNode` and `_add_graph_input_if_need` invocations, as documented in comments update just now ~